datafusion/common/src/diagnostic.rs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use sqlparser::tokenizer::Span;
+
+/// Additional contextual information intended for end users, to help them
+/// understand what went wrong by providing human-readable messages, and
+/// locations in the source query that relate to the error in some way.
+///
+/// You can think of a single [`Diagnostic`] as a single "block" of output from
+/// rustc. i.e. either an error or a warning, optionally with some notes and
+/// help messages.
+///
+/// If the diagnostic, a note, or a help message doesn't need to point to a
+/// specific location in the original SQL query (or the [`Span`] is not
+/// available), use [`Span::empty`].
+///
+/// Example:
+///
+/// ```rust
+/// # use datafusion_common::Diagnostic;
+/// # use sqlparser::tokenizer::Span;
+/// # let span = Span::empty();
+/// let diagnostic = Diagnostic::new_error("Something went wrong", span)
+///     .with_help("Have you tried turning it on and off again?", Span::empty());
+/// ```
+#[derive(Debug, Clone)]
+pub struct Diagnostic {
+    pub kind: DiagnosticKind,
+    pub message: String,
+    pub span: Span,
```

Comment:
Since we don't always have a `Span` I think this would make more sense to be 

```suggestion
    pub span: Option<Span>,
```

Which would make it easier to tell when we didn't have any Span information (and would likely avoid the use of `first_or_empty`)

Reply:
Yeah I agree with you that `Option` is nicer, but I wanted to avoid having two separate "null" values. We'd still have to check for spans that are equal to `Span::EMPTY` somewhere along the chain, because that's what sqlparser returns.

But I'll make the suggested change, let's see how it looks.

Reply:
So after this suggestion and the one about adding a `Span` to `datafusion-common`, I ended up designing it so that a `Span` is always a known location. And then there's a `Span::try_from_sqlparser_span(span: sqlparser::tokenizer::Span) -> Option<Span>` function to do the conversion.

I think it's a lot easier to reason about the lack of information when it's `Option<Span>` and not `sqlparser::tokenizer::Span::EMPTY`